torchdata/stateful_dataloader/README.md

```diff
@@ -0,0 +1,215 @@
+
+for num_workers in [0, 2]:
+  print(f"{num_workers=}")
+  dl = StatefulDataLoader(NoisyRange(5, 1, 1), sampler=MySampler(5, 1, 10),
+      batch_size=2, drop_last=False, num_workers=num_workers)
+
+  batches = []
+  for i, batch in enumerate(dl):
+    batches.append(batch)
+    if i == 2:
+      sd = dl.state_dict()
+
+  dl.load_state_dict(sd)
+  batches2 = list(dl)
+
+  print(batches[3:])
+  print(batches2)
+
+"""
+Output:
+num_workers=0
+[tensor([-0.4526,  3.7948], dtype=torch.float64), tensor([6.5494, 3.0470], dtype=torch.float64)]
+[tensor([-0.4526,  3.7948], dtype=torch.float64), tensor([6.5494, 3.0470], dtype=torch.float64)]
+num_workers=2
+[tensor([3.7412, 1.2438], dtype=torch.float64), tensor([4.4807, 4.0036], dtype=torch.float64)]
+[tensor([3.7412, 1.2438], dtype=torch.float64), tensor([4.4807, 4.0036], dtype=torch.float64)]
+"""
+```
+
+## Saving Custom State with Iterable-Style Datasets
+
+Tracking iteration order with Iterable-style datasets requires state from each worker-level instance of the dataset to
+be captured. You can define `state_dict/load_state_dict` methods on your dataset which capture worker-level state.
+`StatefulDataLoader` will handle aggregation across workers and distribution back to the workers. Calling
+`load_state_dict` requires `StatefulDataLoader` to have same `num_workers` as those of the provided `state_dict`.
+
+```py
+from typing import *
+import torch
+import torch.utils.data
+from torchdata.stateful_dataloader import StatefulDataLoader
+
+
+class MyIterableDataset(torch.utils.data.IterableDataset):
+  def __init__(self, high: int, seed: int):
+    self.high, self.seed = high, seed
+    self.g = torch.Generator()
+    self.i = 0
+
+  def __iter__(self):
+    worker_info = torch.utils.data.get_worker_info()
+    if worker_info is not None:
+      worker_id = worker_info.id
+      num_workers = worker_info.num_workers
+    else:
+      worker_id = 0
+      num_workers = 1
+    self.g.manual_seed(self.seed)
+    arr = torch.randperm(self.high, generator=self.g)
+    arr = arr[worker_id:self.high:num_workers]
+    for idx in range(self.i, len(arr)):
+      self.i += 1
+      yield arr[idx]
+    self.i = 0
+
+  def state_dict(self):
+    return {"i": self.i}
+
+  def load_state_dict(self, state_dict):
+    self.i = state_dict["i"]
+
+for num_workers in [0, 2]:
+  print(f"{num_workers=}")
+  dl = StatefulDataLoader(
+      MyIterableDataset(12, 0), batch_size=2, drop_last=False,
+      num_workers=num_workers)
+
+  batches = []
+  for i, batch in enumerate(dl):
+    batches.append(batch)
+    if i == 2:
+      sd = dl.state_dict()
+
+  dl.load_state_dict(sd)
+  batches2 = list(dl)
+
+  print(batches[3:])
+  print(batches2)
+
+"""
+Output:
+num_workers=0
+[tensor([ 2, 10]), tensor([3, 1]), tensor([11,  6])]
+[tensor([ 2, 10]), tensor([3, 1]), tensor([11,  6])]
+num_workers=2
+[tensor([ 4, 10]), tensor([ 3, 11]), tensor([1, 6])]
+[tensor([ 4, 10]), tensor([ 3, 11]), tensor([1, 6])]
+"""
+```
+
```

Comment:
[This can be done in a separate PR too] Can we also list the corner cases where it doesn't work as expected? Basically this case where user has to pass in an explicit generator for it work correctly - https://github.com/pytorch/data/blob/main/test/stateful_dataloader/test_state_dict.py#L686.

Reply:
Yes that's a good idea. If we could detect and raise assertion errors it would be even better